doc/source/tools.rst

```diff
@@ -1,30 +1,259 @@
+keeps things more stabl.
+
+Deployments
+***********
+
+A deployment is a collection of pods on kubernetes. It is how kubernetes
+knows exactly what containers and what machines need to be running at all
+times. For example, if you have two pods: one that does the authenticating
+described above, and another that manages a database, you can specify both
+in a deployment.
+
+Kubernetes will ensure that both pods are active, and if
+one goes down then it will try to re-create it. It does this by continually
+checking the current state of the pods, and then comparing this with the
+original specification of the deployment. If there are differences between
+the current state vs. the specification of the deployment, Kubernetes will
+attempt to make changes until the current state matches the specification.
+
+**Note**: Users don't generally "create" deployments directly, they are
+instead generated from a set of instructions that are sent to Kubernetes.
+We'll cover this in the section on "Helm".
+
+Service
+*******
+
+A service is simply a stable way of referring to a deployment. Kubernetes
+is all about intelligently handling dynamic and quickly-changing
+computational environments. This means that the hardware may change,
+IP addresses will be different, etc. However you don't want to have to
+re-orient yourself every time this happens. A Kubernetes service keeps
+track of all these changes on the backend, and provides a single address
+to manage your deployment.
+
+Namespace
+*********
+
+Finally, a `namespace <https://kubernetes.io/docs/admin/namespaces/>`_
+defines a collection of deployments. This will define
+the boundaries of hardware that the deployments have to exist within. It
+is generally the most "high-level" of the groups we've discussed thus far.
+For example, a a namespace could be a single class running with JupyterHub.
+It defines all the hardware that is available to students. It also defines
+the JupyterHub machinery that glues together all of the student containers,
+manages disk space, etc.
+
+Persistent Volume Chain
+***********************
+
+Persistent Volume Chains are a way to have persistent storage without
+being tied down to one specific computer or machine. Kubernetes is
+about that flexibility, and that means that we don't want to lock ourselves
+in to a particular operating system just because our files are already
+on it. Persistent Volume Chains help deal with this problem by knowing
+how to convert files between filesystem types. 
+
+
+Helm
+----
+
+`Helm <https://helm.sh/>`_ is a way of specifying kubernetes objects
+with a standard template.
+
+Charts
+******
+
+The way that Helm controls kubernetes is with templates of structured
+information that specify some computational requirements.
+These templates are called "charts", or "helm charts". They contain
+all of the necessary information for kubernetes to generate either:
+
+- a deployment object 
```

Comment:
we generate a bunch of other stuff too, and we generate many deployments and services as well.